### ISSUE TITLE
Implementation details about linear evaluation

### ISSUE THREAD

**Issue by MrChenFeng** [KyleZheng1997/ReSSL#8]
Thanks for your great work and code sharing!
I downloaded the code and logs for small-sized dataset but can't reproduce the linear evaluation accuracy in table 1 following page 6 settings, can you provide some detailed hyperparameter settings? If I'm not wrong, we use 100 epochs linear eval after 200 epochs pertaining?

**Comment by mingkai-zheng** [comment]
we have provided the 800 epochs checkpoints and also the logs for both pretraining and linear evaluation, please see script/batch_train.sh which provides all the scripts that you will need. 

BTW, change the learning rate to 1 and using cosine lr_scheduler for linear evaluation should give a slightly better result in 800 epochs setting.

**Comment by MrChenFeng** [comment]
Hi,

Im actually trying to reproduce the table 1 with cifar datasets. I want to confirm do you use 200 epochs pretrained model for this table? Cause I rerun the pertaining code for 200 epochs several times but only get ~61% for cifar100, with cosine lr-scheduler in eval phase, with linear it's even worse. I'm not here to attack, I just want to know if there are any details that I missed.
![image](https://user-images.githubusercontent.com/24207392/149119641-846a103f-0212-498a-a6d8-9fefae1e750a.png)

**Comment by mingkai-zheng** [comment]
Yes, we pretrain our model for 200 epochs for the result in this table, the pretraining setting should be similar to the one that we provided in script/batch_train.sh, did you follow exactly the same training recipe ?

**Comment by MrChenFeng** [comment]
I used the code from the provided zip.
For hyperparamters:
batchsize: 256
lr: 0.06 with 5 epoch linear warmup + 195 epoch cosine annealing
weak augment for memory encoder while strong augment for student encoder
0.04 for teacher while 0.1 for student as temperature
weight decay as 5e-4
backbone resnet18(modify first conv and remove maxpool), with a two-layer mlp as projector [512-2048-128].

Did I miss something?

**Comment by mingkai-zheng** [comment]
It seems that all the hyperparameters are okay, I'm not sure where is the performance gap comes from. Actually, for the Cifar experiments, you don't need to change any lines of the code, the default configuration is already the best one. Anyway, do you sure you use exactly the same training recipe that we provide? If you are 100% sure about it, I would like to run my code again for 200 epochs to see where is the performance gap.

**Comment by MrChenFeng** [comment]
I think only removed several lines for tinyimagnet for the zipped code which require the memcache things just to ensure the code can be compiled and run cause I don't have the mc package.
That will be very nice if you can run it and upload the logs and code for it. Thanks so much!

**Comment by mingkai-zheng** [comment]
I have run the code again for 200 epochs, it seems that there is no problem with my code.



Here is a brief summary of my pretraining log:
```
Epoch: [0][194/195]	Time  0.130 ( 0.140)	Data  0.000 ( 0.005)	CE 8.1760e+00 (8.1564e+00)
Epoch: [20][194/195]	Time  0.132 ( 0.136)	Data  0.000 ( 0.004)	CE 5.8580e+00 (5.8574e+00)
Epoch: [40][194/195]	Time  0.134 ( 0.136)	Data  0.000 ( 0.004)	CE 5.3201e+00 (5.3725e+00)
Epoch: [60][194/195]	Time  0.132 ( 0.136)	Data  0.000 ( 0.004)	CE 5.1424e+00 (5.0807e+00)
Epoch: [80][194/195]	Time  0.133 ( 0.136)	Data  0.000 ( 0.004)	CE 4.8142e+00 (4.8861e+00)
Epoch: [100][194/195]	Time  0.132 ( 0.136)	Data  0.000 ( 0.004)	CE 4.8314e+00 (4.7264e+00)
Epoch: [120][194/195]	Time  0.131 ( 0.136)	Data  0.000 ( 0.004)	CE 4.3667e+00 (4.5657e+00)
Epoch: [140][194/195]	Time  0.132 ( 0.136)	Data  0.000 ( 0.004)	CE 4.3896e+00 (4.4014e+00)
Epoch: [160][194/195]	Time  0.132 ( 0.136)	Data  0.000 ( 0.004)	CE 4.3312e+00 (4.2264e+00)
Epoch: [180][194/195]	Time  0.133 ( 0.136)	Data  0.000 ( 0.004)	CE 4.2606e+00 (4.0680e+00)
Epoch: [199][194/195]	Time  0.133 ( 0.136)	Data  0.000 ( 0.004)	CE 4.0241e+00 (4.0143e+00)
```

Here is a brief summary of my linear evaluation log (lr=10, step lr_scheduler ):
```
Epoch:0 * Acc@1 57.000 Acc@5 85.990 Best_Acc@1 57.000 Best_Acc@5 85.990
Epoch:10 * Acc@1 60.700 Acc@5 88.060 Best_Acc@1 61.310 Best_Acc@5 88.060
Epoch:20 * Acc@1 60.840 Acc@5 88.080 Best_Acc@1 61.900 Best_Acc@5 88.460
Epoch:30 * Acc@1 61.490 Acc@5 88.420 Best_Acc@1 62.310 Best_Acc@5 88.720
Epoch:40 * Acc@1 61.520 Acc@5 87.930 Best_Acc@1 62.310 Best_Acc@5 88.720
Epoch:50 * Acc@1 61.710 Acc@5 87.870 Best_Acc@1 62.310 Best_Acc@5 88.720
Epoch:60 * Acc@1 63.430 Acc@5 88.840 Best_Acc@1 63.430 Best_Acc@5 88.840
Epoch:70 * Acc@1 63.670 Acc@5 89.060 Best_Acc@1 63.770 Best_Acc@5 89.080
Epoch:80 * Acc@1 63.810 Acc@5 89.080 Best_Acc@1 63.830 Best_Acc@5 89.080
Epoch:90 * Acc@1 64.030 Acc@5 89.070 Best_Acc@1 64.080 Best_Acc@5 89.080
Epoch:99 * Acc@1 64.020 Acc@5 89.090 Best_Acc@1 64.170 Best_Acc@5 89.090
```

With lr=30 and step lr_scheduler, we get 
```
Epoch:99 * Acc@1 63.110 Acc@5 88.330 Best_Acc@1 63.450 Best_Acc@5 88.470
```

With lr=1 and cosine lr_scheduler, we get 
```
Epoch:99 * Acc@1 63.240 Acc@5 89.240 Best_Acc@1 63.510 Best_Acc@5 89.360
```

**Comment by MrChenFeng** [comment]
Thanks so much.
I guess lr matters, in the original code, cos lr in eval was set to 0.2. Let me try it again with lr = 1. How's the knn ACC after 200 pertaining?  In my run, knn ACC reach ~57.4.

**Comment by MrChenFeng** [comment]
Just mentioned it's still open for knn evaluation, sorry.

**Comment by mingkai-zheng** [comment]
The lr=0.2 is for 800 epochs pertaining (actually lr=1 is slightly better for 800 epochs). As we have mentioned in the paper (Section4, Evaluation Protocol part) we use lr=30 and step lr_scheduler by default for the result in Table 1.  We do have the KNN evaluation result for Cifar (but only for 800 epochs), please refer to the open discussions about this paper in [here](https://openreview.net/forum?id=ErivP29kYnx&noteId=cOw3WijRCa) (Re Q2: small number of training epochs).

**Comment by MrChenFeng** [comment]
thanks so much!
